### PR TITLE
Attach Respondent into to Case on Answer Filed

### PR DESCRIPTION
**Pull request by codyseibert**
When a answer is filed, it will attach a "respondent" property to the case:

```javascript
{
  "createdAt": "2018-12-11T22:38:00.796Z",
  "petitionerName": "Test Taxpayer",
  "documents": [
    {
      "createdAt": "2018-12-11T22:38:31.673Z",
      "documentId": "5e2018b4-2189-4b99-a1c2-a5d33845fdf0",
      "documentType": "Petition",
      "userId": "taxpayer",
      "filedBy": "Test Taxpayer"
    },
    {
      "createdAt": "2018-12-11T22:38:31.673Z",
      "documentId": "f7f760f9-f291-446a-9878-2921938abc36",
      "documentType": "Request for Place of Trial",
      "userId": "taxpayer",
      "filedBy": "Test Taxpayer"
    },
    {
      "createdAt": "2018-12-11T22:38:31.673Z",
      "documentId": "c9c26e7a-18cc-48f6-979f-65d6e7dff65d",
      "documentType": "Statement of Taxpayer Identification Number",
      "userId": "taxpayer",
      "filedBy": "Test Taxpayer"
    },
    {
      "createdAt": "2018-12-11T22:38:31.673Z",
      "documentId": "f008b9c9-8971-4346-92a6-2d5b28234bdb",
      "documentType": "Answer",
      "userId": "respondent",
      "filedBy": "Test Respondent"
    }
  ],
  "caseId": "958892fa-a1ac-4d68-a3f8-de592ed0dcef",
  "respondent": {
    "firstName": "Test",
    "lastName": "Respondent",
    "address": "1341 Mesa Drive, Las Vegas, NV, 89119",
    "barNumber": "12345",
    "isIRSAttorney": true,
    "phone": "111-111-1111",
    "title": "Mr",
    "userId": "respondent",
    "email": "respondent@exampe.com"
  },
  "userId": "taxpayer",
  "docketNumber": "110-18",
  "status": "new"
}
```

The respondent info comes from the User.js.

### CODE REVIEW

File: shared/src/business/useCases/respondent/fileAnswer.interactor.test.js [hunk 1]
```diff
@@ -45,11 +50,17 @@ describe('fileAnswer', () => {
       },
       applicationContext,
     });
-    expect(updateCaseStub.getCall(0).args[0].caseDetails).to.contain({
-      respondentId: 'respondent',
-      respondentFirstName: 'bob',
-      respondentLastName: 'marley',
-      respondentBarNumber: '12345',
+    expect(updateCaseStub.getCall(0).args[0].caseDetails.respondent).toEqual({
```
Comment (kkoskelin) [inline shared/src/business/useCases/respondent/fileAnswer.interactor.test.js]: could you simply compare to MOCK_USER?

**Review by kkoskelin — APPROVED** [review]

File: shared/src/business/entities/Case.js [hunk 2]
```diff
@@ -58,10 +59,18 @@ joiValidationDecorator(
       .string()
       .regex(docketNumberMatcher)
       .required(),
-    respondentId: joi.string().optional(),
-    respondentFirstName: joi.string().optional(),
-    respondentLastName: joi.string().optional(),
-    respondentBarNumber: joi.string().optional(),
+    respondent: joi.object().keys({
```
Comment (dknesek) [inline shared/src/business/entities/Case.js:62]: Why wouldn't we create a "Respondent" entity that has this responsibility? That is, when the case it created it plucks this part of the JSON out and creates a new Respondent that validates this. The Case holds a reference to an instance of the respondent entity.

As an aside but related comment, the domain model seems too flat here. I feel like we should have Case --> Answer <-- Repondent.

File: shared/src/business/useCases/respondent/associateRespondentToCase.interactor.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  if (!isAuthorized(userId, FILE_ANSWER)) {
+    throw new UnauthorizedError('Unauthorized for getCasesByStatus');
+  }
+
+  const caseToAssociate = await applicationContext.getUseCases().getCase({
+    caseId,
+    userId,
+    applicationContext,
+  });
+  new Case(caseToAssociate).validate();
```
Comment (dknesek) [inline shared/src/business/useCases/respondent/associateRespondentToCase.interactor.js]: Is there any reason to think getCase() could return an invalid case?

File: shared/src/business/useCases/respondent/associateRespondentToCase.interactor.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+  const user = await applicationContext.getUseCases().getUser(userId);
```
Comment (dknesek) [inline shared/src/business/useCases/respondent/associateRespondentToCase.interactor.js]: At some point, the currently logged in user should be accessible directly from the application context `applicationContext.getUser()`

File: shared/src/business/useCases/respondent/associateRespondentToCase.interactor.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      address: user.address,
+      isIRSAttorney: user.isIRSAttorney,
+      phone: user.phone,
+      barNumber: user.barNumber,
+    },
+  })
+    .validate()
+    .toJSON();
+
+  await applicationContext.getUseCases().updateCase({
```
Comment (dknesek) [inline shared/src/business/useCases/respondent/associateRespondentToCase.interactor.js]: We are calling `updateCase` here. I would expect this to just call a persistence method `getPersistenceGateway().associateAnswerToCase(caseId, answerData, respondentData)` that knows how an answer and respondent manifest in Dynamo. This keeps `updateCase` from becoming this super-method that knows all the nuances of updating case related data in every situation. Makes sense to me to have more targeted persistence functions.

File: shared/src/business/useCases/respondent/associateRespondentToCase.interactor.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    caseId: rawCaseWithRespondent.caseId,
+    caseJson: rawCaseWithRespondent,
+    userId,
+    applicationContext,
+  });
+
+  await applicationContext.getPersistenceGateway().createRespondentCaseMapping({
```
Comment (dknesek) [inline shared/src/business/useCases/respondent/associateRespondentToCase.interactor.js]: This "mapping" would be built into the persistence method I described in the previous comment.

File: shared/src/business/useCases/respondent/fileAnswer.interactor.js [hunk 7]
```diff
@@ -28,29 +28,30 @@ exports.fileAnswer = async ({
   const answerDocumentMetadata = {
```
Comment (dknesek) [inline shared/src/business/useCases/respondent/fileAnswer.interactor.js]: It feels like we've complicated things here. I would expect one use case interactor to be called after the S3 upload, `AssociateAnswerToCase`, that manages all the rest. I think our treating "use cases" as function feels like its right, but it's going to create a chatty and messy situation.

**Review by dknesek — COMMENTED** [review]
This is great work, but I think it exposed some issues we need to iron out.

File: shared/src/business/useCases/respondent/fileAnswer.interactor.js (hunk 7 above)
Comment (codyseibert) [reply shared/src/business/useCases/respondent/fileAnswer.interactor.js]: so maybe AssociateAnswerToCase should hit an endpoint POST@/v1/cases/:caseId/answer and have that endpoint hook everything together?

File: shared/src/business/useCases/respondent/associateRespondentToCase.interactor.js (hunk 3 above)
Comment (codyseibert) [reply shared/src/business/useCases/respondent/associateRespondentToCase.interactor.js]: I guess not if we are always validating what is written.  Are you saying we can just remove this?

**Review by dknesek — APPROVED** [review]

File: shared/src/business/entities/Case.js [hunk 8]
```diff
@@ -58,10 +59,18 @@ joiValidationDecorator(
...
+      userId: joi.string().optional(),
```
Comment (sadlerw) [inline shared/src/business/entities/Case.js]: respondentId might be a better name here

**Review by sadlerw — CHANGES_REQUESTED** [review]

File: shared/src/business/entities/Case.js (hunk 8 above)
Comment (codyseibert) [reply shared/src/business/entities/Case.js]: fixed

**Pull request closed without merging**